modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,54 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
```

Comment:
Pull this callback out into a function that captures intent. Like `hasContentMutations` or something that returns a boolean from an array of mutations.